File: nodejs/src/__tests__/serialize.test.ts
import { describe, expect, it } from "vitest"
import {
  bigintReplacer,
  quoteFromJSON,
  quoteToJSON,
  swapResultFromJSON,
  swapResultToJSON,
  tokenInfoFromJSON,
  tokenInfoToJSON,
} from "../serialize.js"
import type { Quote, SwapResult, TokenInfo } from "../types.js"

const sampleQuote: Quote = {
  tokenIn:      "0x833589fcd6edb6e08f4c7c32d4f71b54bda02913",
  tokenOut:     "0x4200000000000000000000000000000000000006",
  amountIn:     "1000",
  amountOut:    "0.5",
  minOut:       "0.495",
  amountInWei:  1_000_000_000n,
  amountOutWei: 500_000_000_000_000_000n,
  minOutWei:    495_000_000_000_000_000n,
  steps:        "0xdeadbeef",
  path:         ["0x833589fcd6edb6e08f4c7c32d4f71b54bda02913", "0x4200000000000000000000000000000000000006"],
  hops: [{
    tokenIn:      "0x833589fcd6edb6e08f4c7c32d4f71b54bda02913",
    tokenOut:     "0x4200000000000000000000000000000000000006",
    amountIn:     "1000",
    amountOut:    "0.5",
    minOut:       "0.495",
    amountInWei:  1_000_000_000n,
    amountOutWei: 500_000_000_000_000_000n,
    minOutWei:    495_000_000_000_000_000n,
    tokenInPrice:  "1",
    tokenOutPrice: "1",
    slippage:     0.5,
    type:         "v3",
    kind:         "uniswap",
    routeId:      1,
    weight:       1.0,
  }],
  slippage:      0.5,
  feeBps:        35,
  tokenInPrice:  "1",
  tokenOutPrice: "1",
  createdAt:     1_700_000_000_000,
  network:       "base",
  maxHops:       2,
}

describe("bigintReplacer", () => {
  it("converts bigints to strings inside JSON.stringify", () => {
    const out = JSON.stringify({ a: 100n, b: "x", c: 2 }, bigintReplacer)
    expect(out).toBe('{"a":"100","b":"x","c":2}')
  })
})

describe("Quote serialize roundtrip", () => {
  it("preserves all bigints across JSON", () => {
    const json = quoteToJSON(sampleQuote)
    // Ensure no bigints remain — JSON.stringify will throw if any do
    const str = JSON.stringify(json)
    expect(typeof json.amountInWei).toBe("string")
    expect(json.amountInWei).toBe("1000000000")

    const parsed = JSON.parse(str)
    const restored = quoteFromJSON(parsed)
    expect(restored.amountInWei).toBe(sampleQuote.amountInWei)
    expect(restored.minOutWei).toBe(sampleQuote.minOutWei)
    expect(restored.hops[0].amountOutWei).toBe(sampleQuote.hops[0].amountOutWei)
    expect(restored.createdAt).toBe(sampleQuote.createdAt)
  })

  it("accepts a JSON string directly", () => {
    const json = JSON.stringify(quoteToJSON(sampleQuote))
    const restored = quoteFromJSON(json)
    expect(restored.amountInWei).toBe(sampleQuote.amountInWei)
  })
})

describe("SwapResult serialize roundtrip", () => {
  const sample: SwapResult = {
    txHash:      "0xabc",
    blockNumber: 1234567890123n,
    amountIn:    1_000_000_000n,
    amountOut:   500_000_000_000_000_000n,
    tokenIn:     "0x833589fcd6edb6e08f4c7c32d4f71b54bda02913",
    tokenOut:    "0x4200000000000000000000000000000000000006",
    gasUsed:     150_000n,
    effectiveGasPrice: 1_500_000_000n,
    feeWei:      150_000n * 1_500_000_000n,
    feeEth:      "0.000225",
  }
  it("preserves all bigints", () => {
    const json = swapResultToJSON(sample)
    expect(json.blockNumber).toBe("1234567890123")
    const restored = swapResultFromJSON(JSON.stringify(json))
    expect(restored.amountIn).toBe(sample.amountIn)
    expect(restored.amountOut).toBe(sample.amountOut)
    expect(restored.blockNumber).toBe(sample.blockNumber)
    expect(restored.gasUsed).toBe(sample.gasUsed)
  })
})

describe("Quote serialize edge cases", () => {
  it("fromJSON tolerates missing path/hops", () => {
    const minimal = {
      ...quoteToJSON(sampleQuote),
      path: undefined,
      hops: undefined,
    } as any
    const r = quoteFromJSON(minimal)
    expect(r.path).toEqual([])
    expect(r.hops).toEqual([])
  })
})

describe("SwapResult serialize edge cases", () => {
  it("fromJSON tolerates missing fee fields", () => {
    const json = {
      txHash: "0xabc", blockNumber: "1", amountIn: "10", amountOut: "20",
      tokenIn: "0xaa", tokenOut: "0xbb", gasUsed: "100",
      // effectiveGasPrice + feeWei missing
    } as any
    const r = swapResultFromJSON(json)
    expect(r.effectiveGasPrice).toBe(0n)
    expect(r.feeWei).toBe(0n)
  })
})

describe("TokenInfo serialize roundtrip", () => {
  it("handles tokens with balance + allowance", () => {
    const info: TokenInfo = {
      address: "0x833589fcd6edb6e08f4c7c32d4f71b54bda02913",
      symbol:  "USDC",
      name:    "USD Coin",
      decimals: 6,
      owner:   "0x1234567890123456789012345678901234567890",
      balance:   1_000_000n,
      allowance: 500_000n,
    }
    const json = tokenInfoToJSON(info)
    expect(json.balance).toBe("1000000")
    const restored = tokenInfoFromJSON(JSON.stringify(json))
    expect(restored.balance).toBe(info.balance)
    expect(restored.allowance).toBe(info.allowance)
  })

  it("handles metadata-only tokens", () => {
    const info: TokenInfo = { address: "0xa", symbol: "X", name: "X", decimals: 18 }
    const restored = tokenInfoFromJSON(tokenInfoToJSON(info))
    expect(restored.balance).toBeUndefined()
    expect(restored.allowance).toBeUndefined()
  })
})

File: go/serialize_test.go
package afi

import (
	"encoding/json"
	"math/big"
	"testing"

	"github.com/ethereum/go-ethereum/common"
)

func sampleQuote() *Quote {
	return &Quote{
		TokenIn:      common.HexToAddress("0x833589fcd6edb6e08f4c7c32d4f71b54bda02913"),
		TokenOut:     common.HexToAddress("0x4200000000000000000000000000000000000006"),
		AmountIn:     "1000",
		AmountOut:    "0.5",
		MinOut:       "0.495",
		AmountInWei:  big.NewInt(1_000_000_000),
		AmountOutWei: new(big.Int).SetUint64(500_000_000_000_000_000),
		MinOutWei:    new(big.Int).SetUint64(495_000_000_000_000_000),
		Steps:        []byte{0xde, 0xad, 0xbe, 0xef},
		Path: []common.Address{
			common.HexToAddress("0x833589fcd6edb6e08f4c7c32d4f71b54bda02913"),
			common.HexToAddress("0x4200000000000000000000000000000000000006"),
		},
		Hops: []Hop{{
			TokenIn:      common.HexToAddress("0x833589fcd6edb6e08f4c7c32d4f71b54bda02913"),
			TokenOut:     common.HexToAddress("0x4200000000000000000000000000000000000006"),
			AmountIn:     "1000",
			AmountOut:    "0.5",
			MinOut:       "0.495",
			AmountInWei:  big.NewInt(1_000_000_000),
			AmountOutWei: new(big.Int).SetUint64(500_000_000_000_000_000),
			MinOutWei:    new(big.Int).SetUint64(495_000_000_000_000_000),
			TokenInPrice: "1", TokenOutPrice: "1",
			Slippage: 0.5, Type: "v3", Kind: "uniswap", RouteID: 1, Weight: 1.0,
		}},
		Slippage:      0.5,
		FeeBps:        35,
		TokenInPrice:  "1",
		TokenOutPrice: "1",
		CreatedAt:     1_700_000_000_000,
	}
}

func TestQuote_MarshalJSON_RoundTrip(t *testing.T) {
	q := sampleQuote()
	data, err := json.Marshal(q)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}

	// Wei fields should be quoted strings, not raw numbers
	if want := `"amountInWei":"1000000000"`; !contains(string(data), want) {
		t.Errorf("expected %q in JSON, got: %s", want, data)
	}

	var restored Quote
	if err := json.Unmarshal(data, &restored); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if restored.AmountInWei.Cmp(q.AmountInWei) != 0 {
		t.Errorf("AmountInWei mismatch: %s vs %s", restored.AmountInWei, q.AmountInWei)
	}
	if restored.MinOutWei.Cmp(q.MinOutWei) != 0 {
		t.Errorf("MinOutWei mismatch: %s vs %s", restored.MinOutWei, q.MinOutWei)
	}
	if restored.Hops[0].AmountOutWei.Cmp(q.Hops[0].AmountOutWei) != 0 {
		t.Error("Hop amountOutWei mismatch")
	}
	if restored.CreatedAt != q.CreatedAt {
		t.Errorf("CreatedAt mismatch: %d vs %d", restored.CreatedAt, q.CreatedAt)
	}
}

func TestSwapResult_MarshalJSON_RoundTrip(t *testing.T) {
	r := &SwapResult{
		TxHash:      common.HexToHash("0xabc"),
		BlockNumber: 1234567890123,
		AmountIn:    big.NewInt(1_000_000_000),
		AmountOut:   new(big.Int).SetUint64(500_000_000_000_000_000),
		TokenIn:     common.HexToAddress("0x833589fcd6edb6e08f4c7c32d4f71b54bda02913"),
		TokenOut:    common.HexToAddress("0x4200000000000000000000000000000000000006"),
		GasUsed:     150_000,
	}
	data, err := json.Marshal(r)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	if !contains(string(data), `"amountIn":"1000000000"`) {
		t.Errorf("amountIn should be string-encoded: %s", data)
	}

	var restored SwapResult
	if err := json.Unmarshal(data, &restored); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if restored.AmountIn.Cmp(r.AmountIn) != 0 {
		t.Error("AmountIn mismatch")
	}
	if restored.BlockNumber != r.BlockNumber {
		t.Error("BlockNumber mismatch")
	}
}

func TestTokenInfo_MarshalJSON_WithBalance(t *testing.T) {
	info := &TokenInfo{
		Address:   common.HexToAddress("0x833589fcd6edb6e08f4c7c32d4f71b54bda02913"),
		Symbol:    "USDC",
		Name:      "USD Coin",
		Decimals:  6,
		Owner:     common.HexToAddress("0x1234567890123456789012345678901234567890"),
		Balance:   big.NewInt(1_000_000),
		Allowance: big.NewInt(500_000),
	}
	data, err := json.Marshal(info)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	if !contains(string(data), `"balance":"1000000"`) {
		t.Errorf("balance string-encoded missing: %s", data)
	}

	var restored TokenInfo
	if err := json.Unmarshal(data, &restored); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if restored.Balance.Cmp(info.Balance) != 0 {
		t.Error("Balance mismatch")
	}
}

func TestTokenInfo_MetadataOnly(t *testing.T) {
	info := &TokenInfo{
		Address:  common.HexToAddress("0xabc"),
		Symbol:   "X",
		Name:     "X",
		Decimals: 18,
	}
	data, _ := json.Marshal(info)
	var restored TokenInfo
	if err := json.Unmarshal(data, &restored); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if restored.Balance != nil {
		t.Errorf("Balance should be nil, got %v", restored.Balance)
	}
}

func contains(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}
